### PR TITLE
Check for matching ConnectionEvent

### PR DESCRIPTION
**Pull request by yschimke**
Refactor `CallEvent.closes` and take same approach for ConnectionEvent.

### CODE REVIEW

File: okhttp-testing-support/src/main/kotlin/okhttp3/OkHttpClientTestRule.kt [hunk 1]
```diff
@@ -173,7 +173,9 @@ class OkHttpClientTestRule : BeforeEachCallback, AfterEachCallback {
       // a test timeout failure.
       val waitTime = (entryTime + 1_000_000_000L - System.nanoTime())
       if (!queue.idleLatch().await(waitTime, TimeUnit.NANOSECONDS)) {
-        TaskRunner.INSTANCE.cancelAll()
+        synchronized (TaskRunner.INSTANCE) {
```
Comment (yschimke) [inline okhttp-testing-support/src/main/kotlin/okhttp3/OkHttpClientTestRule.kt:176]: The fail wasn't happening, because cancelAll fails.

Comment (swankjesse) [reply okhttp-testing-support/src/main/kotlin/okhttp3/OkHttpClientTestRule.kt:176]: oooh good catch

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt [hunk 2]
```diff
@@ -106,15 +114,34 @@ open class RecordingConnectionListener : ConnectionListener() {
         .isFalse()
     }
 
+    if (enforceOrder) {
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:117]: Nice.

**Review by swankjesse — APPROVED** [review]
This is a net improvement, but it doesn’t check as much as we could.

In particular, this confirms every close has a matching open, but not that every open has a matching close.

**Comment by swankjesse** [comment]
(I think we probably need a `@Rule` for that!)

**Pull request merged**